File: behaviorial/visitor/typescript/concrete-visitor.ts
import { Invoice, Resume, Report } from './concrete-element';
import { DocumentVisitor } from './visitor';

export class PrintVisitor implements DocumentVisitor {
	visitReport(report: Report): void {
		report.generateReport();
		console.log('Printing report.');
	}

	visitInvoice(invoice: Invoice): void {
		invoice.createInvoice();
		console.log('Printing invoice.');
	}

	visitResume(resume: Resume): void {
		resume.prepareResume();
		console.log('Printing resume.');
	}
}

export class SaveToDatabaseVisitor implements DocumentVisitor {
	visitReport(report: Report): void {
		report.generateReport();
		console.log('Saving report to database.');
	}

	visitInvoice(invoice: Invoice): void {
		invoice.createInvoice();
		console.log('Saving invoice to database.');
	}

	visitResume(resume: Resume): void {
		resume.prepareResume();
		console.log('Saving resume to database.');
	}
}

File: behaviorial/visitor/python/concrete_visitor.py
from concrete_element import Invoice, Report, Resume
from visitor import DocumentVisitor


class PrintVisitor(DocumentVisitor):
    def visit_report(self, report: Report):
        report.generate_content()
        print("Printing report:")

    def visit_invoice(self, invoice: Invoice):
        invoice.generate_content()
        print("Printing invoice:")

    def visit_resume(self, resume: Resume):
        resume.generate_content()
        print("Printing resume:")


class SaveToDatabaseVisitor(DocumentVisitor):
    def visit_report(self, report: Report):
        report.generate_content()
        print("Saving report to database:")

    def visit_invoice(self, invoice: Invoice):
        invoice.generate_content()
        print("Saving invoice to database:")

    def visit_resume(self, resume: Resume):
        resume.generate_content()
        print("Saving resume to database:")
